form/form.go

```diff
@@ -0,0 +1,279 @@
+	t := val.Type()
+
+	// Also do nothing if this is the type's zero value
+	if t.Comparable() && val.Interface() == reflect.Zero(t).Interface() {
+		return
+	}
+
+	switch val.Kind() {
+	case reflect.Array, reflect.Slice:
+		// formatName automatically adds square brackets, so just pass an empty
+		// string into the breadcrumb trail
+		arrNames := append(names, "")
+
+		for i := 0; i < val.Len(); i++ {
+			// The one exception to the above is when options have requested
+			// that this array/slice be indexed. In that case we produce a hash
+			// keyed with integers which the Stripe API knows how to interpret.
+			if options != nil && options.IndexedArray {
+				arrNames = append(names, strconv.Itoa(i))
+			}
+
+			reflectValue(values, val.Index(i), arrNames, nil)
+		}
+
+	case reflect.Bool:
+		if options != nil {
+			if val.Bool() {
+				switch {
+				case options.Empty:
+					values.Add(formatName(names), "")
+				case options.Invert:
+					values.Add(formatName(names), strconv.FormatBool(false))
+				case options.Zero:
+					values.Add(formatName(names), "0")
+				}
+			}
+		} else {
+			values.Add(formatName(names), strconv.FormatBool(val.Bool()))
+		}
+
+	case reflect.Float32:
+		values.Add(formatName(names), strconv.FormatFloat(val.Float(), 'f', 4, 32))
+
+	case reflect.Float64:
+		values.Add(formatName(names), strconv.FormatFloat(val.Float(), 'f', 4, 64))
+
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		values.Add(formatName(names), strconv.FormatInt(val.Int(), 10))
+
+	case reflect.Interface:
+		if val.IsNil() {
+			return
+		}
+		reflectValue(values, val.Elem(), names, nil)
+
+	case reflect.Map:
+		for _, keyVal := range val.MapKeys() {
+			if keyVal.Kind() != reflect.String {
+				panic("Don't support serializing maps with non-string keys")
+			}
+
+			reflectValue(values, val.MapIndex(keyVal), append(names, keyVal.String()), nil)
+		}
+
+	case reflect.Ptr:
+		if val.IsNil() {
+			return
+		}
+		reflectValue(values, val.Elem(), names, options)
+
+	case reflect.String:
+		values.Add(formatName(names), val.String())
+
+	case reflect.Struct:
+		for i := 0; i < t.NumField(); i++ {
+			field := t.Field(i)
+			tag := field.Tag.Get(tagName)
+			if tag == "" {
+				continue
+			}
+			formName, options := parseTag(tag)
+
+			if options != nil &&
+				(options.Empty || options.Invert || options.Zero) &&
+				val.Field(i).Type().Kind() != reflect.Bool {
+
+				panic("Cannot specify `zero` on non-integer field")
+			}
+
+			reflectValue(values, val.Field(i), append(names, formName), options)
+		}
+
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
+		values.Add(formatName(names), strconv.FormatUint(val.Uint(), 10))
+	}
+}
+
+func parseTag(tag string) (string, *formOptions) {
+	var options *formOptions
+	parts := strings.Split(tag, ",")
+	name := parts[0]
+
+	for i := 1; i < len(parts); i++ {
+		switch parts[i] {
+		case "empty":
```

Comment:
@remi-stripe Yeah. Only one of them would be honored though because the combination isn't conceptually sound.